### ISSUE TITLE
Not thread safe

### ISSUE THREAD

**Issue by alikoncak** [ArXen42/SerilogSuppressLoggingExample#1]
It seems like that this code is not thread safe.

I was intenting to use it on .net6.

I created an alternative you might consider.

```
public sealed class SerilogExtensions: IDisposable
  {
    public static AsyncLocal<bool> LoggingSuppressed = new();

    /// <summary>
    /// Constructor suppresses logging
    /// </summary>
    public SerilogExtensions()
    {
      LoggingSuppressed.Value = true;
    }

    /// <summary>
    /// Not actual dispose, but just disable suppression
    /// </summary>
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CA1816:Dispose methods should call SuppressFinalize", Justification = "Not actual disposing something")]
    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1063:Implement IDisposable Correctly", Justification = "Not actual disposing something")]
    public void Dispose()
    {
      LoggingSuppressed.Value = false;
    }
  }
```

**Comment by ArXen42** [comment]
Thanks for mentioning it. I haven't updated this example in a long time, probably should do that, current `HashSet` solution is indeed not ideal. Looking at this code now I'm not even sure why I used it in the first place, it seems redundant.

However, I'm not sure if `AsyncLocal` is needed. Here I use [LogContext](https://github.com/serilog/serilog/blob/dev/src/Serilog/Context/LogContext.cs) to push `"SuppressLogging"` marker property. `LogContext` already uses `AsyncLocal` under the hood as seen in it's source, so only logging code inside one `async` call chain should be affected.

In my own code where I use this I encountered a problem with logging not being fully suppressed when suppressed code is spawning new threads, which do not share the same `AsyncLocal` ambient context.

So for the project with this problem I plan to just introduce explicit `ILogger` dependency for everything that needs logging, because some parts became too complex to rely on `AsyncLocal` ambient context anyway (suppressed modules spawn their own worker threads, use `ConfigureAwait(false)`, etc, that breaks `AsyncLocal` context).

To summarize, when I created this sample and corresponding question on stackoverflow, I logged everything using global logger like `Log.Information`. Now I consider this to be problematic approach and prefer to inject `ILogger` instances, so for each application module it may contain customized properties, including suppression if needed.

I will try to find time to look into potential problems and update this repo for both ambient `LogContext` and more explicit `ILogger` approach.
